### PR TITLE
ugrep - normalize CRLF output

### PR DESCRIPTION
**Pull request by idigdoug**
README.md:
- Update the README to reflect that the GLOB argument does work but that
  `-g` is still likely to be more intuitive.

input.cpp:
- Don't change file mode in file_init. Assume file is binary. (This also
  means the `enc` parameter is no longer needed.)

output.hpp:
- Convert CRLF and CR to LF before writing text to an output file.

query.cpp:
- Open pipe as binary.

ugrep.cpp:
- Open searched files and pipes as binary.
- Update usage for the -U/--binary parameter.

### CODE REVIEW

**Comment by genivia-inc** [comment]
Looks in principle fine and follows what I had in mind to change the `fwrite()` in output.cpp and disabling the text/binary parts for Windows that isn't useful (I admit I'm not a big fan of C/C++ Windows programming with its archaic rules and legacy "modernized" API).

Converting the output buffer might not be the most efficient way to replace CRLF with LF, because the loop cannot be vectorized, but it can be unrolled. Perhaps `memchr()` that is usually highly optimized can be used to find the next `\n` and then check for `\r\n` pair to output the line with `fwrite()` up to a replaced `\r` with `\n` is more efficient (if lines are not too short).

This assumes that the output is always text mode, which I presume is the case even when redirecting the output to a file or to a pipe (yes?).

**Comment by idigdoug** [comment]
memchr is likely a bit faster than this loop, but each call to memchr and fwrite has some overhead. The overhead would be proportional to the number of CR bytes. On Windows, where CR is common, I would expect the overhead to be greater than the savings. On Unix, where CR is uncommon, I would expect the savings to be greater.

memchr + memmove would also work. There is startup overhead for both of them. When CR is common, memchr + memmove might be slightly faster or slightly slower -- I would have to benchmark. When CR is rare, memchr + memmove would likely be faster.

I suspect that it won't make a substantial difference in overall ugrep performance since the loop is already pretty efficient and the output is usually going to be a small faction of the total ugrep processing time. But I'd be willing to change this to a memchr + memmove implementation.

**Comment by genivia-inc** [comment]
Thanks for the changes.

Two comments:
1) Removing CR should only apply to Windows OS (to avoid double CRCRLF) thus leaving the output the same on other OS. Ideally the output should reflect the matched input.

2) The logic currently removes all CR, not limited to CRLF and CRCRLF. To remove CR from CRLF/CRCRLF is a bit more of a hassle since the CR could be at the end of a block with the next block starting with LF. So a flag/count is needed to record the fact that a block ended in one or more CR, which is/are then not output by fwrite(). The next block outputs LF without the CRs or outputs CR if the block does not start with LF. This method is more precise since it doesn't remove a singular CR, but may be overkill perhaps.

**Comment by idigdoug** [comment]
1. I don't agree that CR in the input should be propagated to the output. ugrep's output is "lines that match". ugrep's definition of a line is the dominating concern. Adding extra CR to the output messes up the output without adding useful information.
2. The original commit implemented the logic you describe. As mentioned in the discussion for the issue, it did not work well. The most important issue was that console control codes get inserted before LF, which turns each CRLF into an isolated CR followed by control codes followed by isolated LF. I changed to logic that simply removes all CR.

**Comment by genivia-inc** [comment]
You may or may not agree with me or anyone else, but the fact is that GNU/BSD grep preserves all CR in the input:
~~~
echo $'hello\x0djoe' | grep ''
joelo
~~~
The output should be a faithful reproduction of the input when all lines are matched. We don't want to introduce surprises that somehow and for no reason the output differs from the input. Normalizing CR and LF is best delegated to `tr`, `sed` and `awk`.

Further:
1) The logic I had suggested reads everything in binary, which is happening but the logic isn't correct.
2) It is a debatable technical point whether to output in text or binary. If binary, then it's automatically reproducing the input without modification. If the output is text, the CRCRLF becomes apparent. The extra CR in CRCRLF should be avoided.

**Comment by idigdoug** [comment]
I can respect that, but it's different from what we discussed earlier. I wish this preference had been expressed earlier. 

I'm not sure that the behavior of BSD/GNU grep is the model for ideal tool behavior. It is the behavior you get when you don't give any thought to line endings. I'm not convinced that it is the behavior you actually want in a cross-platform world where line endings are not consistent.

1. What is incorrect about the PR's logic that reads everything in binary?
2. We have previously discussed the tradeoffs of outputting in text or binary.
  - If we want to emulate BSD/GNU grep then yes, ugrep should output in binary. It is then outputting "chunks of the input" rather than "lines". We accept that the consistency of ugrep's output will depend on the consistency of ugrep's input.
  - If we want to have ugrep generate nicely formatted output then it should use consistent line endings. It is then outputting "lines" rather than "chunks of the input". The simplest way to achieve this is to remove all CR and output in text mode. (Simple removal of CR is consistent with the rest of ugrep's behavior because ugrep does not treat isolated CR as a line-ending.)

In the issue discussion, I stated several times that would be ok with binary output, but that I would much prefer consistent output. I asked if you had a preference and didn't really get a response, so I assumed you were ok with an implementation of consistent output.

If you are certain this is your preference then I'll make another iteration of this PR with binary output.

**Comment by idigdoug** [comment]
I've implemented binary output. It works, but it's a very ugly change.

There are so many places where ugrep outputs `chr('\n')` or `nl()`. Each of these has to be evaluated and fixed. Some of them remain `'\n'`. Some of them are changed into `NEWLINE` (platform-specific macro). And some of them are `\n` or `NEWLINE` depending on what happened immediately before.

I've probably missed some cases or made some errors, but it seems to work ok.

If this is the direction you want to go, I'll push a commit with this change.

**Comment by idigdoug** [comment]
I've updated the PR to use binary mode for output.

I don't like this solution. It means that ugrep maintainers will always have to be careful about writing the correct line ending. There are also cases where ugrep will output inconsistent line endings by-design, and users will be confused about why and about what they need to do to fix things.

But it seems to work ok in my limited testing.

**Comment by genivia-inc** [comment]
Thanks! Will take a closer look, a bit preoccupied at the moment. Because this does not affect input, there should not be any impact on users' writing the correct line ending (your quote). If there is a concern, it's on a different topic. The only difference I would expect is in the output to avoid getting an extra CR for the CRLF pairs.

**Comment by genivia-inc** [comment]
I've reviewed the changes in the source files. The changes look fine to me. It is the right approach IMO. I'll look into this over the weekend to check it our more thoroughly after merging.

**Review by genivia-inc — APPROVED** [review]

**Comment by genivia-inc** [comment]
@idigdoug the latest release v3.6 adds two files `matcher_avx2.cpp` and `matcher_avx512bw.cpp` that are both identical to `matcher.cpp` (symlinked in my internal repo). These are compiled with different `-m` arch options to force AVX2 and AVX512BW, respectively, to produce `simd_match_avx2()` and `simd_match_avx512bw()`. For ugrep.exe bundled with the ugrep repo in bin/win32 and bin/win64, only `matcher_avx2.cpp` with AVX2 is compiled, so `matcher_avx512bw.cpp` is not used at all (until perhaps a later update).

These changes may break your PR patch for Windows builds, so I wanted to give you a heads-up on that. The part I've added to the vs/ugrep/README.md is:
~~~
Right-click on the matcher_avx2.cpp file to set the C++ preprocessor properties specifically for this file by adding COMPILE_AVX2 to the list of preprocessor definitions, and enable AVX2:

        C/C++
                Preprocessor: 
                        Preprocessor Definitions: WIN32;NDEBUG;_CONSOLE;WITH_NO_INDENT;HAVE_AVX2;COMPILE_AVX2;HAVE_PCRE2;PCRE2_STATIC;HAVE_LIBZ;HAVE_LIBBZ2;HAVE_LIBLZMA;HAVE_LIBLZ4;HAVE_LIBZSTD;WITH_COLOR;ZLIB_WINAPI;NO_GZCOMPRESS;LZMA_API_STATIC;_CRT_NONSTDC_NO_DEPRECATE;_CRT_SECURE_NO_DEPRECATE;_CRT_NONSTDC_NO_WARNINGS
                Code Generation:
                        Enable Enhanced Instruction Set: Advanced Vector Extensions 2 (/arch:AVX2) 

If the general C++ preprocessor properties and the C++ preprocessor properties for matcher_avx2.cpp are the same, then linkage fails with duplicate definition errors.

Remove matcher_avx512bw.cpp from the list of source files.

Then build ugrep in Visual Studio.
~~~
The reason for all this is explained in the [v3.6.0 release note](https://github.com/Genivia/ugrep/releases/tag/v3.6.0).

**Comment by idigdoug** [comment]
I’ll take a look at how to best handle this in the msvc project

In situations like this, I usually make a .INL file with the code, and then make separate .CPP files that #include the .INL file, or I pick one .cpp to be the “base” file. That allows the git repo to work the same even when symlinks aren’t propagated, as is the case with git repos.

In addition, this provides the opportunity to perform configuration within the .CPP file before including the .INL or base .CPP file.

For example, matcher_avx2 might be:

#if !defined(__AVX2__) || defined(__AVX512BW__)
#error matcher_avx2.cpp must be compiled with /arch:avx2.
#endif
#define COMPILE_AVX2
#include "matcher.cpp”

I strongly prefer to have as much configuration as possible in the code (.H or .CPP) rather than in project files (makefile or vcxproj). Everything that has to be in “Preprocessor Definitions” is a hidden burden, easy to accidentally mess up.

Thanks!
Doug

**Pull request merged**
